File: Fort.ind UWP/Services/SitemapService.vb
Imports System.Xml.Linq
Imports Windows.Storage

''' <summary>
''' Parses the bundled sitemap.xml and produces SearchItem entries for every URL
''' </summary>
Public Class SitemapService

    ''' <summary>
    ''' Reads sitemap.xml from the app package and returns SearchItem objects allowing for the latest URLs to be searchable 
    ''' </summary>
    Public Shared Async Function LoadSearchItemsAsync() As Task(Of List(Of SearchItem))
        Dim items As New List(Of SearchItem)

        Try
            Dim cachedUrls = Await TryLoadCachedUrlsAsync()
            If cachedUrls IsNot Nothing AndAlso cachedUrls.Count > 0 Then
                Return BuildSearchItemsFromUrls(cachedUrls)
            End If

            Dim file = Await StorageFile.GetFileFromApplicationUriAsync(New Uri("ms-appx:///sitemap.xml"))
            Dim text = Await FileIO.ReadTextAsync(file)
            
            ' Protect against malformed XML
            Dim doc As XDocument = Nothing
            Try
                doc = XDocument.Parse(text)
            Catch xmlEx As Exception
                Debug.WriteLine($"SitemapService: XML parsing failed – {xmlEx.Message}")
                Return items ' Return empty list if XML is malformed
            End Try
            
            Dim ns As XNamespace = "http://www.sitemaps.org/schemas/sitemap/0.9"
            Dim urlsToCache As New List(Of String)()

            For Each urlElement In doc.Descendants(ns + "url")
                Dim urlValue = urlElement.Element(ns + "loc")?.Value
                If String.IsNullOrEmpty(urlValue) Then Continue For

                urlsToCache.Add(urlValue)

                Dim item = CreateSearchItemFromUrl(urlValue)
                If item IsNot Nothing Then
                    items.Add(item)
                End If
            Next

            If urlsToCache.Count > 0 Then
                Await SaveCachedUrlsAsync(urlsToCache)
            End If
        Catch ex As Exception
            Debug.WriteLine($"SitemapService: failed to load sitemap – {ex.Message}")
        End Try

        Return items
    End Function

    ''' <summary>
    ''' Creates a SearchItem instance from a URL string, or returns Nothing if the URL
    ''' is invalid or should be skipped (e.g. utility pages like 404).
    ''' </summary>
    ''' <param name="urlValue">The absolute URL string.</param>
    Private Shared Function CreateSearchItemFromUrl(urlValue As String) As SearchItem
        If String.IsNullOrWhiteSpace(urlValue) Then
            Return Nothing
        End If

        Dim uri As Uri = Nothing
        If Not Uri.TryCreate(urlValue, UriKind.Absolute, uri) Then
            Return Nothing
        End If

        Dim path = uri.AbsolutePath.Trim("/"c)
        If String.IsNullOrEmpty(path) Then
            Return New SearchItem("Home", AppConstants.CategoryFortWebsite, Nothing, urlValue)
        End If

        ' Skip utility pages
        If path = "404" Then
            Return Nothing
        End If

        Dim category = GetCategory(path)
        Dim title = GetTitle(path)
        Return New SearchItem(title, category, Nothing, urlValue)
    End Function

    Private Shared Function BuildSearchItemsFromUrls(urls As IEnumerable(Of String)) As List(Of SearchItem)
        Dim items As New List(Of SearchItem)()

        For Each urlValue In urls
            Dim item = CreateSearchItemFromUrl(urlValue)
            If item IsNot Nothing Then
                items.Add(item)
            End If
        Next

        Return items
    End Function

    Private Shared Async Function TryLoadCachedUrlsAsync() As Task(Of List(Of String))
        Try
            Dim settings = ApplicationData.Current.LocalSettings
            If Not settings.Values.ContainsKey(AppConstants.SitemapCacheTimestampKey) Then
                Return Nothing
            End If

            Dim rawTimestamp = settings.Values(AppConstants.SitemapCacheTimestampKey)
            Dim cacheUnixSeconds As Long
            Try
                cacheUnixSeconds = Convert.ToInt64(rawTimestamp)
            Catch ex As FormatException
                Return Nothing
            Catch ex As InvalidCastException
                Return Nothing
            Catch ex As OverflowException
                Return Nothing
            End Try

            Dim nowUnixSeconds = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
            Dim maxAgeSeconds = CLng(AppConstants.SitemapCacheTtlHours) * 60L * 60L
            If (nowUnixSeconds - cacheUnixSeconds) > maxAgeSeconds Then
                Return Nothing
            End If

            Dim cacheFile = Await ApplicationData.Current.LocalFolder.GetFileAsync(AppConstants.SitemapCacheFileName)
            Dim content = Await FileIO.ReadTextAsync(cacheFile)
            If String.IsNullOrWhiteSpace(content) Then
                Return Nothing
            End If

            Dim urls As New List(Of String)()
            Dim lines = content.Split({vbCrLf, vbLf}, StringSplitOptions.RemoveEmptyEntries)
            For Each line In lines
                Dim value = line.Trim()
                If Not String.IsNullOrWhiteSpace(value) Then
                    urls.Add(value)
                End If
            Next

            If urls.Count = 0 Then
                Return Nothing
            End If

            Return urls
        Catch ex As Exception
            Debug.WriteLine($"SitemapService: failed to load sitemap cache – {ex.Message}")
            Return Nothing
        End Try
    End Function

    Private Shared Async Function SaveCachedUrlsAsync(urls As IEnumerable(Of String)) As Task
        Try
            Dim lines As New List(Of String)()
            For Each url In urls
                If Not String.IsNullOrWhiteSpace(url) Then
                    lines.Add(url)
                End If
            Next

            If lines.Count = 0 Then
                Return
            End If

            Dim cacheFile = Await ApplicationData.Current.LocalFolder.CreateFileAsync(
                AppConstants.SitemapCacheFileName,
                CreationCollisionOption.ReplaceExisting)

            Await FileIO.WriteTextAsync(cacheFile, String.Join(Environment.NewLine, lines))
            ApplicationData.Current.LocalSettings.Values(AppConstants.SitemapCacheTimestampKey) = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        Catch ex As Exception
            Debug.WriteLine($"SitemapService: failed to save sitemap cache – {ex.Message}")
        End Try
    End Function

    Private Shared Function GetCategory(path As String) As String
        If path.StartsWith("games/html/") Then Return "Games — HTML"
        If path.StartsWith("games/flash/") Then Return "Games — Flash"
        If path.StartsWith("games/codepen/") Then Return "Games — CodePen"
        If path.StartsWith("games/retroclassic-mostly-emulated/") Then Return "Games — Retro"
        If path.StartsWith("games/minecraft/") Then Return "Games — Minecraft"
        If path.StartsWith("games/") Then Return "Games"
        If path.StartsWith("social/") Then Return "Social"
        If path.StartsWith("emulators/") Then Return "Emulators"
        If path.StartsWith("apps/appstone/") Then Return "Apps — AppStone"
        If path.StartsWith("apps/") Then Return "Apps"
        If path.StartsWith("extras/") Then Return "Extras"
        If path.StartsWith("labs-betas/") Then Return "Labs & Betas"
        Return AppConstants.CategoryFortWebsite
    End Function

    Private Shared Function GetTitle(path As String) As String
        ' Use the last segment of the path as the display name when showing results! example: "games/html/rynis-game" -> "Rynis Game"
        Dim trimmed = path.TrimEnd("/"c)
        Dim lastSlash = trimmed.LastIndexOf("/"c)
        Dim slug = If(lastSlash >= 0, trimmed.Substring(lastSlash + 1), trimmed)

        If String.IsNullOrEmpty(slug) Then Return path

        ' Title-case in a single pass with one StringBuilder
        Dim sb As New System.Text.StringBuilder(slug.Length)
        Dim capitalizeNext As Boolean = True
        For i = 0 To slug.Length - 1
            Dim c = slug(i)
            If c = "-"c OrElse c = "_"c Then
                sb.Append(" "c)
                capitalizeNext = True
            ElseIf capitalizeNext Then
                sb.Append(Char.ToUpper(c))
                capitalizeNext = False
            Else
                sb.Append(Char.ToLower(c))
            End If
        Next
        Return sb.ToString()
    End Function

End Class

File: Fort.ind UWP/Services/SitemapService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Windows.Storage;

namespace Fort.ind_UWP
{
    public class SitemapService
    {
        private static volatile IReadOnlyList<SearchItem> s_allItems;
        private static volatile IReadOnlyList<SearchItem> s_gameItems;

        private static readonly SemaphoreSlim s_loadGate = new SemaphoreSlim(1, 1);

        public static async Task<IReadOnlyList<SearchItem>> LoadSearchItemsAsync()
        {
            var cached = s_allItems;
            if (cached != null) return cached;

            await s_loadGate.WaitAsync();
            try
            {
                if (s_allItems != null) return s_allItems;

                var parsed = await ParseSitemapAsync();
                var result = parsed.AsReadOnly();
                if (parsed.Count > 0)
                {
                    s_allItems = result;
                }
                return result;
            }
            finally
            {
                s_loadGate.Release();
            }
        }

        public static async Task<IReadOnlyList<SearchItem>> LoadGameItemsAsync()
        {
            var cachedGames = s_gameItems;
            if (cachedGames != null) return cachedGames;

            var all = await LoadSearchItemsAsync();

            var result = all.Where(item => item.CategoryKey != null &&
                                           item.CategoryKey.StartsWith(AppConstants.CategoryGames, StringComparison.Ordinal))
                            .ToList()
                            .AsReadOnly();
            if (all.Count > 0)
            {
                s_gameItems = result;
            }
            return result;
        }

        private static async Task<List<SearchItem>> ParseSitemapAsync()
        {
            List<SearchItem> items = new List<SearchItem>();

            try
            {
                var file = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///Assets/sitemap.xml"));
                var text = await FileIO.ReadTextAsync(file);

                var entries = await Task.Run(() => ReadSitemapEntries(text));
                if (entries == null)
                {
                    return items;
                }

                items = BuildSearchItems(entries);

                DeleteLegacyUrlCacheInBackground();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"SitemapService: failed to load sitemap – {ex.Message}");
            }

            return items;
        }

        private static List<string> ReadLocValues(string documentText)
        {
            List<string> urls = new List<string>();

            System.Xml.XmlReaderSettings settings = new System.Xml.XmlReaderSettings()
            {
                DtdProcessing = System.Xml.DtdProcessing.Prohibit,
                XmlResolver = null,
                IgnoreComments = true,
                IgnoreWhitespace = true,
                IgnoreProcessingInstructions = true
            };

            try
            {
                using (System.IO.StringReader stringReader = new System.IO.StringReader(documentText))
                {
                    using (var reader = System.Xml.XmlReader.Create(stringReader, settings))
                    {
                        while (reader.Read())
                        {
                            if (reader.NodeType == System.Xml.XmlNodeType.Element &&
                                string.Equals(reader.LocalName, "loc", StringComparison.Ordinal))
                            {
                                var value = reader.ReadElementContentAsString();
                                if (!string.IsNullOrWhiteSpace(value))
                                {
                                    urls.Add(value.Trim());
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"SitemapService: XML parsing failed – {ex.Message}");
                return null;
            }

            return urls;
        }

        private sealed class SitemapEntry
        {
            public string Title;
            public string CategoryKey;
            public string Url;
        }

        private static List<SitemapEntry> ReadSitemapEntries(string documentText)
        {
            var urls = ReadLocValues(documentText);
            if (urls == null)
            {
                return null;
            }

            return urls.Select(CreateEntryFromUrl)
                       .Where(entry => entry != null)
                       .ToList();
        }

        private static SitemapEntry CreateEntryFromUrl(string urlValue)
        {
            var uri = WebLauncher.TryCreateWebUri(urlValue);
            if (uri == null)
            {
                return null;
            }

            var path = uri.AbsolutePath.Trim('/');
            if (string.IsNullOrEmpty(path))
            {
                return new SitemapEntry { Title = null, CategoryKey = AppConstants.CategoryFortWebsite, Url = urlValue };
            }

            if (path == "404")
            {
                return null;
            }

            return new SitemapEntry { Title = GetTitle(path), CategoryKey = GetCategory(path), Url = urlValue };
        }

        private static List<SearchItem> BuildSearchItems(List<SitemapEntry> entries)
        {
            List<SearchItem> items = new List<SearchItem>(entries.Count);
            foreach (var entry in entries)
            {
                var title = entry.Title ?? LocalizedStrings.Get("SearchItemHome");
                items.Add(new SearchItem(title, entry.CategoryKey, null, entry.Url));
            }
            return items;
        }

        private static async void DeleteLegacyUrlCacheInBackground()
        {
            try
            {
                var values = ApplicationData.Current.LocalSettings.Values;
                if (!values.ContainsKey(AppConstants.LegacySitemapCacheTimestampKey)) return;

                var cacheFile = await ApplicationData.Current.LocalFolder.TryGetItemAsync(AppConstants.LegacySitemapCacheFileName);
                if (cacheFile != null)
                {
                    await cacheFile.DeleteAsync(StorageDeleteOption.PermanentDelete);
                }

                values.Remove(AppConstants.LegacySitemapCacheTimestampKey);
                values.Remove(AppConstants.LegacySitemapCacheAppVersionKey);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"SitemapService: could not remove the legacy URL cache – {ex.Message}");
            }
        }

        private static string GetCategory(string path)
        {
            if (path.StartsWith("games/html/", StringComparison.Ordinal)) return AppConstants.CategoryGamesHtml;
            if (path.StartsWith("games/flash/", StringComparison.Ordinal)) return AppConstants.CategoryGamesFlash;
            if (path.StartsWith("games/codepen/", StringComparison.Ordinal)) return AppConstants.CategoryGamesCodePen;
            if (path.StartsWith("games/retroclassic-mostly-emulated/", StringComparison.Ordinal)) return AppConstants.CategoryGamesRetro;
            if (path.StartsWith("games/minecraft/", StringComparison.Ordinal)) return AppConstants.CategoryGamesMinecraft;
            if (path.StartsWith("games/", StringComparison.Ordinal)) return AppConstants.CategoryGames;
            if (path.StartsWith("social/", StringComparison.Ordinal)) return AppConstants.CategorySocial;
            if (path.StartsWith("emulators/", StringComparison.Ordinal)) return AppConstants.CategoryEmulators;
            if (path.StartsWith("apps/appstone/", StringComparison.Ordinal)) return AppConstants.CategoryAppsAppStone;
            if (path.StartsWith("apps/", StringComparison.Ordinal)) return AppConstants.CategoryApps;
            if (path.StartsWith("extras/", StringComparison.Ordinal)) return AppConstants.CategoryExtras;
            if (path.StartsWith("labs-betas/", StringComparison.Ordinal)) return AppConstants.CategoryLabsAndBetas;
            return AppConstants.CategoryFortWebsite;
        }

        private static readonly HashSet<string> s_upperCaseTokens = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "cs", "css", "dbz", "fnaf", "gba", "gbc", "gta", "hd", "html", "mlb", "mlg",
            "n64", "nba", "nds", "nes", "nfl", "nhl", "psp", "snes", "tmnt", "tv", "ufc",
            "ufo", "wwe"
        };

        private static readonly HashSet<string> s_domainSuffixTokens = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "com", "gg", "io", "lol", "net", "org"
        };

        private static string GetTitle(string path)
        {
            var trimmed = path.TrimEnd('/');
            var lastSlash = trimmed.LastIndexOf('/');
            var slug = lastSlash >= 0 ? trimmed.Substring(lastSlash + 1) : trimmed;

            if (string.IsNullOrEmpty(slug)) return path;

            var tokens = slug.Split(new[] { '-', '_' }, StringSplitOptions.RemoveEmptyEntries);
            if (tokens.Length == 0) return path;

            System.Text.StringBuilder sb = new System.Text.StringBuilder(slug.Length);
            for (int i = 0; i <= tokens.Length - 1; i++)
            {
                var token = tokens[i];

                if (i > 0 && s_domainSuffixTokens.Contains(token))
                {
                    sb.Append('.');
                    sb.Append(token.ToLowerInvariant());
                    continue;
                }

                if (i > 0 && IsAllDigits(token) && IsAllDigits(tokens[i - 1]))
                {
                    sb.Append('.');
                    sb.Append(token);
                    continue;
                }

                if (i > 0) sb.Append(' ');
                sb.Append(FormatToken(token));
            }

            return sb.ToString();
        }

        private static string FormatToken(string token)
        {
            if (s_upperCaseTokens.Contains(token)) return token.ToUpperInvariant();

            System.Text.StringBuilder sb = new System.Text.StringBuilder(token.Length);
            sb.Append(char.ToUpperInvariant(token[0]));
            for (int i = 1; i <= token.Length - 1; i++)
            {
                sb.Append(char.ToLowerInvariant(token[i]));
            }
            return sb.ToString();
        }

        private static bool IsAllDigits(string token)
        {
            if (string.IsNullOrEmpty(token)) return false;
            for (int i = 0; i <= token.Length - 1; i++)
            {
                if (!char.IsDigit(token[i])) return false;
            }
            return true;
        }
    }
}
